From 379afa723858663a5d351908024d0f5655570364 Mon Sep 17 00:00:00 2001
From: Thomas Epperson <thomas.epperson@gmail.com>
Date: Tue, 8 Oct 2024 11:00:44 -0500
Subject: [PATCH] Add python setup script to try to use module with litex
Upstream-Status: Pending

---
 setup.py | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)
 create mode 100644 setup.py

diff --git a/setup.py b/setup.py
new file mode 100644
index 0000000..b426303
--- /dev/null
+++ b/setup.py
@@ -0,0 +1,34 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+version_str = "1.65"
+
+setuptools.setup(
+    name="pythondata-brianhg-ddr3",
+    version=version_str,
+    author="Brian HG",
+    author_email="BrianHG",
+    description="""\
+Python module containing hdl files for ddr3 controller.""",
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    url="https://github.com/BrianHGinc/BrianHG-DDR3-Controller",
+    classifiers=[
+        "Programming Language :: Python :: 3",
+        "License :: OSI Approved :: MIT License",
+        "Operating System :: OS Independent",
+    ],
+    python_requires='>=3.5',
+    zip_safe=False,
+    packages=setuptools.find_packages(),
+    package_data={
+    	'brianhg-ddr3': ['BrianHG_DDR3/**'],
+    },
+    include_package_data=True,
+    project_urls={
+        "Bug Tracker": "https://github.com/BrianHGinc/BrianHG-DDR3-Controller/issues",
+        "Source Code": "https://github.com/BrianHGinc/BrianHG-DDR3-Controller",
+    },
+)
-- 
2.34.1

